### PR TITLE
 feat(new encoding): add pretty json encoding 

### PR DESCRIPTION
**Pull request by lsampras**
Adds a `use_pretty_json` field to `json` encoding codec
this relies on the `serde_json::to_writer_pretty` implementation while using the same deserialization as plain json

Fixes : #20380 

example
```yaml
sources:
  dummy_logs:
    type: "demo_logs"
    format: "syslog"
    interval: 1
# Parse Syslog logs
# See the Vector Remap Language reference for more info: https://vrl.dev
transforms:
  parse_logs:
    type: "remap"
    inputs: ["dummy_logs"]
    source: |
      . = parse_syslog!(string!(.message))
# Print parsed logs to stdout
sinks:
  print:
    type: "console"
    inputs: ["parse_logs"]
    encoding:
      codec: "json"
      json:
        use_pretty_json: true
```

### CODE REVIEW

**Review by hestonhoffman — APPROVED** [review]

**Comment by lsampras** [comment]
> Also, I'm curious if you considered making this an optional parameter to the existing json codec rather than a separate codec. I think I could go either way on it so just curious if you considered it and what your thoughts were if so.

Hadn't considered adding a parameter to existing json codec, tbh I wasn't aware about the possibilty of adding more fields inside a codec...

**Comment by jszwedko** [comment]
> > Also, I'm curious if you considered making this an optional parameter to the existing json codec rather than a separate codec. I think I could go either way on it so just curious if you considered it and what your thoughts were if so.
> 
> Hadn't considered adding a parameter to existing json codec, tbh I wasn't aware about the possibilty of adding more fields inside a codec...

You can see an example with the AVRO codec: https://vector.dev/docs/reference/configuration/sinks/amqp/#encoding.avro

> @jszwedko Can you take a look at this again? Should I be adding a field in the `json` codec instead of a new codec?

I'm having trouble coming up with super strong arguments either way but I think I have a slight preference to have it just be an option on the existing `json` codec since it is really just controlling "formatting" of an existing encoding. I could see us adding other formatting options to the other existing codecs where having `pretty_json` be its own codec would become more of an outlier. Do you think you that's something you might be able to take on? Apologies for not suggesting it sooner in [the issue you had opened](https://github.com/vectordotdev/vector/issues/20380).

**Comment by pront** [comment]
> > @jszwedko Can you take a look at this again? Should I be adding a field in the `json` codec instead of a new codec?
> 
> I'm having trouble coming up with super strong arguments either way but I think I have a slight preference to have it just be an option on the existing `json` codec since it is really just controlling "formatting" of an existing encoding. I could see us adding other formatting options to the other existing codecs where having `pretty_json` be its own codec would become more of an outlier. Do you think you that's something you might be able to take on? Apologies for not suggesting it sooner in [the issue you had opened](https://github.com/vectordotdev/vector/issues/20380).

+1 on adding a new option (vs introducing a new encoder). 

Happy to help with implementation details if you get stuck. As Jesse mentioned, this https://github.com/vectordotdev/vector/blob/v0.37.1/lib/codecs/src/encoding/format/avro.rs#L42-L52 is a very good example.

File: config/vector.yaml [hunk 1]
```diff
@@ -37,7 +37,9 @@ sinks:
     type: "console"
     inputs: ["parse_logs"]
     encoding:
-      codec: "pretty_json"
+      codec: "json"
+      json:
+        use_pretty_json: true
```
Comment (jszwedko) [inline config/vector.yaml:42]:
```suggestion
        pretty: true
```

I think we can shorten this up a bit.

**Review by jszwedko — COMMENTED** [review]
Looking good! I just have one comment about the option name, but otherwise this looks good to me.

**Review by jszwedko — APPROVED** [review]
Great, thank you @lsampras !

**Comment by lsampras** [comment]
@jszwedko Can you help here? I'm unable to understand the failure
found the following logs in the actions
```
The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.
```

**Comment by jszwedko** [comment]
> @jszwedko Can you help here? I'm unable to understand the failure found the following logs in the actions
> 
> ```
> The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.
> ```

Ah, apologies, this was an issue in CI that was caused due to the Zookeeper and Kafka images disappearing. We resolved that in https://github.com/vectordotdev/vector/commit/3a115c517fe91f4c70eb8211b6dfdd1899adf07c. I'll requeue this.

**Comment by lsampras** [comment]
@jszwedko can you re-queue this for merge?

**Comment by jszwedko** [comment]
Thanks for the bump @lsampras . I merged in master to fix the CI failure and requeued this.

**Pull request merged**